### PR TITLE
rAAVE Implementation

### PR DESCRIPTION
**Pull request by raugfer**
This is AAVE an elastic supply token (similar to AMPL and YAM) that uses AAVE as its reference token.

### CODE REVIEW

File: contracts/GElasticTokenManager.sol [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+			_mintAmount = _totalSupply.mul(_mintPercent).div(1e18);
+		}
+
+		return (_delta, _positive, _mintAmount);
+	}
+
+	function _rebaseAvailable(Self storage _self) internal view returns (bool _available)
+	{
+		if (!_self.rebaseActive) return false;
+		if (now < _self.lastRebaseTime.add(_self.rebaseMinimumInterval)) return false;
```
Comment (ejossev) [inline contracts/GElasticTokenManager.sol:128]: Woudn't it be better to have rather:
```
if (now < _self.lastRebaseTime.add(_self.rebaseMinimumInterval).sub(_self.rebaseWindowLength))
```
? 
The code on the last line suggest that rebase can happen only between 5PM-6PM UTC - but if yesterday it happened e.g. 5:59:50PM yesterday, today window is closed.

File: contracts/GElasticTokenManager.sol [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	{
+		require(_self._rebaseAvailable(), "not available");
+
+		_self.lastRebaseTime = now.sub(now.mod(_self.rebaseMinimumInterval)).add(_self.rebaseWindowOffset);
+		_self.epoch = _self.epoch.add(1);
+
+		_positive = _exchangeRate > 1e18;
+
+		uint256 _deviation = _positive ? _exchangeRate.sub(1e18) : uint256(1e18).sub(_exchangeRate);
+		if (_deviation < _self.rebaseMaximumDeviation) {
```
Comment (ejossev) [inline contracts/GElasticTokenManager.sol:108]: Shouldn't ```_self.rebaseMaximumDeviation``` be called rather ```_self.rebaseMinimumDeviation```?

File: contracts/GElasticTokenManager.sol (hunk 1 above)
Comment (raugfer) [reply contracts/GElasticTokenManager.sol:128]: Yes. I think I saw this on YAM's code base and asked the same question. I will double check to verify.

File: contracts/GElasticTokenManager.sol (hunk 2 above)
Comment (raugfer) [reply contracts/GElasticTokenManager.sol:108]: Makes sense. By rebaseMaximumDeviation I had rebaseMaximumTolerableDeviation in my mind, but rebaseMinimumDeviation makes much more sense.

File: contracts/GElasticTokenManager.sol (hunk 1 above)
Comment (ejossev) [reply contracts/GElasticTokenManager.sol:128]: Any conclusion?

Comment (raugfer) [reply contracts/GElasticTokenManager.sol:128]: They test exact this condition on YAM (https://github.com/yam-finance/yamV3/blob/8a3d82b40b5ab8486b7bdec649049b638d8cacc3/contracts/rebaser/YAMRebaser.sol#L417). I will fix this.

Comment (raugfer) [reply contracts/GElasticTokenManager.sol:128]: Ok, in fact it seems right. The lastRebaseTime always is set to the beginning of the rebaseWindowOffset. and this means that rebaseMinimumInterval is indeed the minimum interval required.

```_self.lastRebaseTime = now.sub(now.mod(_self.rebaseMinimumInterval)).add(_self.rebaseWindowOffset);```

**Review by ejossev — APPROVED** [review]
LGTM

**Comment by raugfer** [comment]
This code has been migrated to its own repository: https://github.com/GrowthDeFi/raave-v1-core

**Pull request closed without merging**